24.0.0.7-beta staging (#3839)
OpenLiberty/blogs: posts/2024-07-02-24.0.0.7-beta.adoc

---
layout: post
title: "Enhancements in Jakarta RESTful Web Services 4.0 and more in 24.0.0.7-beta"
# Do NOT change the categories section
categories: blog
author_picture: https://avatars3.githubusercontent.com/ramkumar-k-9286
author_github: https://github.com/ramkumar-k-9286
seo-title: Enhancements in Jakarta RESTful Web Services 4.0 and more in 24.0.0.7-beta - OpenLiberty.io

> seo-description: The 24.0.0.7-beta release introduces enhancements to Jakarta RESTful Web Services 4.0, including new API methods and media type values. Also, it includes the monitor-1.0 feature for enhanced HTTP request tracking and the beta release of audit-2.0.

Reviewer: ```suggestion
seo-description: The 24.0.0.7-beta release introduces enhancements to Jakarta RESTful Web Services 4.0, including new API methods and media type values. It also includes enhanced HTTP request tracking and the beta release of the Audit 2.0 feature, which creates rest handler records only when a rest handler application is configured.
```

> blog_description: The 24.0.0.7-beta release introduces enhancements to Jakarta RESTful Web Services 4.0, including new API methods and media type values. Also, it includes the monitor-1.0 feature for enhanced HTTP request tracking and the beta release of audit-2.0.

Reviewer: ```suggestion
blog_description: The 24.0.0.7-beta release introduces enhancements to Jakarta RESTful Web Services 4.0, including new API methods and media type values. It also includes enhanced HTTP request tracking and the beta release of the Audit 2.0 feature, which creates rest handler records only when a rest handler application is configured.
```

open-graph-image: https://openliberty.io/img/twitter_card.jpg
open-graph-image-alt: Open Liberty Logo
---
= Enhancements in Jakarta RESTful Web Services 4.0 and more in 24.0.0.7-beta
Ramkumar K <https://github.com/ramkumar-k-9286>
:imagesdir: /
:url-prefix:
:url-about: /
//Blank line here is necessary before starting the body of the post.

> The 24.0.0.7-beta release introduces enhancements to Jakarta RESTful Web Services 4.0, including new API methods and media type values. Also, it includes the monitor-1.0 feature for enhanced HTTP request tracking and the beta release of audit-2.0.

Reviewer: ```suggestion
The 24.0.0.7-beta release introduces enhancements to Jakarta RESTful Web Services 4.0, including new API methods and media type values. It also includes enhanced HTTP request tracking and the beta release of the Audit 2.0 feature, which creates rest handler records only when a rest handler application is configured.
```

* <<webservices, Jakarta Restful Web Services 4.0 (EE11)>>
* <<monitor10, Enhanced HTTP Request Monitoring with Monitor 1.0>>
* <<faces41, Jakarta EE11 Faces 4.1>>
* <<audit20, Use the Audit 2.0 feature to avoid generating unnecessary rest handler records>>

// // // // // // // //
// In the preceding section:
// Change SUB_FEATURE_TITLE to the feature that is included in this release and
// change the SUB_TAG_1/2/3 to the heading tags
//
// However if there's only 1 new feature, delete the previous section and change it to the following sentence:
// "The link:{url-about}[Open Liberty] 24.0.0.7-beta includes SUB_FEATURE_TITLE"
// // // // // // // //

See also link:{url-prefix}/blog/?search=beta&key=tag[previous Open Liberty beta blog posts].

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // // 
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/28707
// Contact/Reviewer: jim-krueger
// // // // // // // // 
[#webservices]
== Jakarta Restful Web Services 4.0 (EE11)

> Jakarta RESTful Web Services provides a foundational API to develop web services based on the REST architectural pattern. The Jakarta Restful Web Services 4.0 update, introduces several enhancements to make web service development more efficient and effective.

Reviewer: ```suggestion
Jakarta RESTful Web Services provides a foundational API to develop web services based on the REST architectural pattern. The Jakarta Restful Web Services 4.0 update introduces several enhancements to make web service development more efficient and effective.
```

> The new features, `restfulWS-4.0` and `restfulWSClient-4.0`, are primarily aimed at application developers. This release brings incremental improvements that are made to Jakarta RESTful Web Services provided in the 4.0 release of that Jakarta project. It provides new convenience API methods:

Reviewer: ```suggestion
The new features, `restfulWS-4.0` and `restfulWSClient-4.0`, are primarily aimed at application developers. This release brings incremental improvements to Jakarta RESTful Web Services. It provides the following new convenience API methods:
```

> - The `containsHeaderString()` method has been added to `ContainerRequestContext`, `ContainerResponseContext`, and `HttpHeaders` to help determine whether a header contains a specified value.

Reviewer: ```suggestion
- The `containsHeaderString()` method is added to `ContainerRequestContext`, `ContainerResponseContext`, and `HttpHeaders` to help determine whether a header contains a specified value.
```

- The `UriInfo.getMatchedResourceTemplate()` method is a new addition that returns a full template for a request.

> - New `MediaType` values have been introduced, including `APPLICATION_JSON_PATCH_JSON` and `APPLICATION_MERGE_PATCH_JSON`, to handle specific media types.

Reviewer: ```suggestion
- New `MediaType` values , to handle specific media types are introduced, including `APPLICATION_JSON_PATCH_JSON` and `APPLICATION_MERGE_PATCH_JSON`.
```

> These updates aim to streamline the process of developing RESTful services. To implement these new features, add the following to your `server.xml` file.

Reviewer: ```suggestion
These updates streamline the process of developing RESTful services. To implement these new features, add the following code to your `server.xml` file.
```

----

For the RESTful web services client feature.
[source,xml]
----
<featureManager>
    <feature>restfulWSClient-4.0</feature>
</featureManager>
----

For more information, see the Jakarta RESTful Web Services 4.0 link:https://jakarta.ee/specifications/restful-ws/4.0/[specification] and the link:https://jakarta.ee/specifications/restful-ws/4.0/apidocs/jakarta.ws.rs/module-summary[Javadoc].

    
// DO NOT MODIFY THIS LINE. </GHA-BLOG-TOPIC> 

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // // 
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/28693
// Contact/Reviewer: Channyboy
// // // // // // // // 
[#monitor10]
== Enhanced HTTP Request Monitoring with Monitor 1.0

Open Liberty introduces a significant update that enhances the tracking of HTTP connections that are made to the server. With the Monitor-1.0 feature (`monitor-1.0`) and any JEE10 feature that uses the servlet engine, Open Liberty can now record detailed information about HTTP requests into an `HttpStatsMXBean`. This includes the request method, response status, duration, HTTP route, and other attributes that align with the Open Telemetry HTTP metric semantic convention.

> If `mpMetrics-5.0` or `mpMetrics-5.1` is also active, Open Liberty creates corresponding HTTP metrics, such as `http.server.request.duration`, using a Timer type metric. To use these metrics effectively, it is essential to configure histogram buckets explicitly for `Timer` and `Histogram` type metrics by using the MicroProfile Metrics. For more details, see the `mp.metrics.distribution.histogram.buckets` property on the link:https://openliberty.io/docs/latest/microprofile-config-properties.html#metrics[MicroProfile Config Properties page].

Reviewer: ```suggestion
If  the `mpMetrics-5.0` or `mpMetrics-5.1` feature is also active, Open Liberty creates corresponding HTTP metrics, such as `http.server.request.duration`, using a Timer type metric. To use these metrics effectively, configure histogram buckets explicitly for `Timer` and `Histogram` type metrics by using the MicroProfile Metrics. For more details, see the `mp.metrics.distribution.timer.buckets` property on the link:https://openliberty.io/docs/latest/microprofile-config-properties.html#metrics[MicroProfile Config Properties page].
```

> If `mpMetrics-5.0` or `mpMetrics-5.1` is also active, Open Liberty creates corresponding HTTP metrics, such as `http.server.request.duration`, using a Timer type metric. To use these metrics effectively, it is essential to configure histogram buckets explicitly for `Timer` and `Histogram` type metrics by using the MicroProfile Metrics. For more details, see the `mp.metrics.distribution.histogram.buckets` property on the link:https://openliberty.io/docs/latest/microprofile-config-properties.html#metrics[MicroProfile Config Properties page].

Reviewer 2: Just only need to configure `Timer` metric type (since the metric is a Timer).
Consequently the property would be the `mp.metrics.distribution.timer.buckets` for the Timer.

I realized i used the  `mp.metrics.distribution.histogram.buckets` one in the issue, sorry!

[source,xml]
----
<featureManager>
    <feature>monitor-1.0</feature>
    <feature>servlet-6.0</feature>
    <feature>mpMetrics-5.0</feature>
</featureManager>
----

For detailed configuration, see the link:https://openliberty.io/docs/latest/microprofile-config-properties.html#metrics[MicroProfile Config Properties page]. 

For more information about the HTTP semantic convention used, see the link:https://opentelemetry.io/docs/specs/semconv/http/http-metrics/#metric-httpserverrequestduration[Open Telemetry HTTP semantic convention] page.
    
    
// DO NOT MODIFY THIS LINE. </GHA-BLOG-TOPIC> 

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // // 
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/28603
// Contact/Reviewer: volosied,pnicolucci
// // // // // // // // 
[#faces41]
== Jakarta EE11 Faces 4.1

> Jakarta Faces is a Model-View-Controller (MVC) framework for building web applications, offering many convenient features such as state management and input validation. The Jakarta Faces 4.1 (`faces-4.1`) update introduces a range of incremental improvements that are aimed at enhancing the developer experience. These enhancements include increased use of generics in FacesMessages and other locations. Improved CDI integration now requires events for `@Initialized`, `@BeforeDestroyed`, and `@Destroyed` for built-in scopes like FlowScoped and ViewScoped, and it allows injecting the current Flow into a backing bean. Also, a new UUIDConverter is added to the API, simplifying the handling of UUIDs.

Reviewer: ```suggestion
Jakarta Faces is a Model-View-Controller (MVC) framework for building web applications, offering many convenient features, such as state management and input validation. The Jakarta Faces 4.1 (`faces-4.1`) update introduces a range of incremental improvements to enhance the developer experience. These enhancements include increased use of generics in `FacesMessages` and other locations. Improved CDI integration by requiring events for `@Initialized`, `@BeforeDestroyed`, and `@Destroyed` for built-in scopes like `FlowScoped` and `ViewScoped`, and it allows injecting the current `Flow` into a backing bean. Also, a new `UUIDConverter` is added to the API, which simplifies the handling of UUIDs.
```

Reviewer 3: +1 for above. 

And can you reword this line? 
`Improved CDI integration now requires events` -> `Improved CDI integration **by requiring** events`

This update also includes some deprecations and removals. The composite: extension tag and references to the Security Manager are removed, and the full state saving mechanism is deprecated. To enable the Jakarta Faces 4.1 feature in your application, add the following to your `server.xml` file:

[source,xml]
----
<featureManager>
    <feature>faces-4.1</feature>
</featureManager>

----

For more information, see the Jakarta Faces 4.1 link:https://jakarta.ee/specifications/faces/4.1/[specification] and the link:https://jakarta.ee/specifications/faces/4.1/apidocs/jakarta.faces/module-summary.html[Javadoc].
    
    
// DO NOT MODIFY THIS LINE. </GHA-BLOG-TOPIC> 

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // // 
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/28557
// Contact/Reviewer: wrodrig
// // // // // // // // 
[#audit20]
== Use the Audit 2.0 feature to avoid generating unnecessary rest handler records

> The Audit 2.0 feature (`audit-2.0`) is now in beta release. The feature is designed for customers that are not using rest handler applications. 

Reviewer: ```suggestion
The Audit 2.0 feature (`audit-2.0`) is now in beta release. The feature is designed for users that are not using rest handler applications. 
```
